File: src/main/java/com/sbna/fxfeed/consumer/converter/MarketPriceConverter.java
package com.sbna.fxfeed.consumer.converter;

import com.sbna.fxfeed.model.MarketPrice;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Locale;

import static java.lang.Long.parseLong;


@Component
public class MarketPriceConverter {

    @Value("${converter.commission.ask}")
    private Double ask;

    @Value("${converter.commission.bid}")
    private Double bid;

    @Value("${converter.delimiter}")
    private String delimiter;

    private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy' 'HH:mm:ss:SSS", Locale.ENGLISH);

    public List<MarketPrice> convert(String message) {
        return message.lines()
                .map(it -> it.split(delimiter))
                .map(it -> new MarketPrice(
                        parseLong(it[0]),
                        it[1].trim(),
                        Double.valueOf(it[2]),
                        Double.valueOf(it[3]),
                        applyCommission(Double.valueOf(it[2]), bid),
                        applyCommission(Double.valueOf(it[3]), ask),
                        LocalDateTime.parse(it[4], formatter)
                )).toList();
    }

    private Double applyCommission(Double price, Double commission) {
        return price + price * commission;
    }
}

File: src/main/kotlin/com/sbna/fxfeed/consumer/converter/MarketPriceConverter.kt
package com.sbna.fxfeed.consumer.converter

import com.sbna.fxfeed.model.MarketPrice
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*


@Component
class MarketPriceConverter {

    @Value("\${converter.commission.ask}")
    private val ask: Double? = null

    @Value("\${converter.commission.bid}")
    private val bid: Double? = null

    @Value("\${converter.delimiter}")
    private val delimiter: String? = null

    private val formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy' 'HH:mm:ss:SSS", Locale.ENGLISH)

    fun convert(message: String): List<MarketPrice> {
        val transformLine: (String) -> List<String> = { it.split(delimiter!!) }
        val transformField: (List<String>) -> MarketPrice = {
            MarketPrice(
                it[0].toLong(),
                it[1],
                it[2].toDouble(),
                it[3].toDouble(),
                applyCommission(it[2].toDouble(), bid),
                applyCommission(it[3].toDouble(), ask),
                LocalDateTime.parse(it[4], formatter)
            )
        }
        return message.lines().map(
            transformLine
        ).map(transformField)
    }

    fun applyCommission(price: Double, commission: Double?): Double {
        return price + price * commission!!
    }
}